ln-gateway/src/cln.rs

```diff
@@ -40,3 +61,233 @@ pub struct HtlcAccepted {
+    Ok(json!({ "address": address }))
+}
+
+async fn deposit_rpc(
+    plugin: Plugin<PluginState>,
+    value: serde_json::Value,
+) -> Result<serde_json::Value, Error> {
+    let deposit: DepositPayload = serde_json::from_value(value)?;
+    let txid = plugin.state().send(deposit).await?;
+    Ok(json!({ "fedimint_txid": txid.to_string() }))
+}
+
+async fn withdraw_rpc(
+    plugin: Plugin<PluginState>,
+    value: serde_json::Value,
+) -> Result<serde_json::Value, Error> {
+    let withdraw: WithdrawPayload = serde_json::from_value(value)?;
+    let txid = plugin.state().send(withdraw).await?;
+    Ok(json!({ "fedimint_txid": txid.to_string() }))
+}
+
+/// Build the core-lightning plugin and register it in an LnRpc
+/// This process generates and writes rpc configs to a known location
+///
+/// Returns
+///
+/// * CLN LnRpc
+/// * The Ln Rpc bind address
+/// * Working directory hosting the rpc config path
+pub async fn build_cln_rpc(
+    sender: GatewayMessageSender,
+) -> Result<(Arc<dyn LnRpc>, SocketAddr, PathBuf), Error> {
+    // Register this plugin with core-lightning
+    if let Some(plugin) = Builder::new(sender, stdin(), stdout())
+        .option(options::ConfigOption::new(
+            "fedimint-cfg",
+            // FIXME: cln_plugin doesn't support parameters without defaults
+            options::Value::String("default-dont-use".into()),
+            "fedimint config directory",
+        ))
+        .option(options::ConfigOption::new(
+            "fedimint-host",
+            options::Value::String("127.0.0.1".into()),
+            "gateway hostname",
+        ))
+        .option(options::ConfigOption::new(
+            "fedimint-port",
+            options::Value::String("8080".into()),
+            "gateway port",
+        ))
+        .rpcmethod("gw-balance", "Display ecash token balance", balance_rpc)
+        .rpcmethod(
+            "gw-deposit",
+            "Deposit into federation. Args: <txoutproof> <bitcoin-transaction>",
+            deposit_rpc,
+        )
+        .rpcmethod(
+            "gw-withdraw",
+            "Withdraw from federation. Args: <address> <sats>",
+            withdraw_rpc,
+        )
+        .rpcmethod("gw-address", "Generate deposit address", address)
+        .hook("htlc_accepted", |plugin, value| async move {
+            // This callback needs to be `Sync`, so we use tokio::spawn
+            let handle = tokio::spawn(async move {
+                htlc_accepted_hook(plugin, value).await.or_else(|e| {
+                    error!("htlc_accepted error {:?}", e);
+                    // cln_plugin doesn't handle errors very well ... tell it to proceed normally
+                    Ok(json!({ "result": "continue" }))
+                })
+            });
+            handle.await?
+        })
+        .dynamic() // Allow reloading the plugin
+        .start()
+        .await?
+    {
+        let workdir = match plugin.option("fedimint-cfg") {
```

Comment:
Well, actually looks like "building and registering" are one thing. Bug generating these configs looks like a candidate for another function.